third_party/amd/lib/TritonAMDGPUToLLVM/TargetInfo.cpp

```diff
@@ -126,10 +126,146 @@ Value TargetInfo::programId(RewriterBase &rewriter, Location loc,
+      reduceOp->getOperand(1) != block->getArgument(1))
+    return std::nullopt;
+
+  if (isa<arith::AddIOp>(reduceOp) || isa<arith::AddFOp>(reduceOp) ||
+      isa<arith::AndIOp>(reduceOp) || isa<arith::MaximumFOp>(reduceOp) ||
+      isa<arith::MaxNumFOp>(reduceOp) || isa<arith::MaxSIOp>(reduceOp) ||
+      isa<arith::MaxUIOp>(reduceOp) || isa<arith::MinimumFOp>(reduceOp) ||
+      isa<arith::MinNumFOp>(reduceOp) || isa<arith::MinSIOp>(reduceOp) ||
+      isa<arith::MinUIOp>(reduceOp) || isa<arith::MulFOp>(reduceOp) ||
+      isa<arith::OrIOp>(reduceOp) || isa<arith::SubFOp>(reduceOp) ||
+      isa<arith::SubIOp>(reduceOp) || isa<arith::XOrIOp>(reduceOp)) {
+    return reduceOp;
+  }
+
+  return std::nullopt;
+}
+
 bool TargetInfo::warpReduce(RewriterBase &rewriter, Location loc,
                             SmallVector<Value> &acc, triton::ReduceOp op,
                             unsigned numLaneToReduce,
                             unsigned interleave) const {
+  if (numLaneToReduce != 64)
+    return false;
+
+  if (auto optionalReduxOp = getReduxOpForDPP(op)) {
+    auto reduxOp = *optionalReduxOp;
+
+    auto createDppReduxOp = [&rewriter, &loc, &op,
+                             &reduxOp](Type valType, Value &src, int dppCtrl,
+                                       int rowMask, int bankMask,
+                                       bool boundCtrl) -> Value {
+      unsigned bits = valType.getIntOrFloatBitWidth();
+
+      Type actualType = valType;
+      if (!valType.isF32()) {
+        if (!valType.isIntOrIndex()) {
+          src = bitcast(src, int_ty(bits));
+          actualType = int_ty(bits);
+        }
+
+        if (bits < 32) {
+          src = sext(i32_ty, src);
+          actualType = i32_ty;
+        }
+      }
+
+      auto dppResult = rewriter
+                           .create<ROCDL::DPPUpdateOp>(
+                               loc, actualType, src, src,
+                               mlir::IntegerAttr::get(i32_ty, dppCtrl),
+                               mlir::IntegerAttr::get(i32_ty, rowMask),
+                               mlir::IntegerAttr::get(i32_ty, bankMask),
+                               mlir::BoolAttr::get(op.getContext(), boundCtrl))
+                           .getRes();
+
+      if (!valType.isF32()) {
+        if (bits < 32) {
+          src = trunc(int_ty(bits), src);
+          dppResult = trunc(int_ty(bits), dppResult);
+        }
+
+        if (!valType.isIntOrIndex()) {
+          src = bitcast(src, valType);
+          dppResult = bitcast(dppResult, valType);
+        }
+      }
+
+      IRMapping mapping;
+      mapping.map(reduxOp->getOperand(0), src);
+      mapping.map(reduxOp->getOperand(1), dppResult);
+      return rewriter.clone(*reduxOp, mapping)->getResult(0);
+    };
+
+    for (int i = 0; i < acc.size(); i++) {
+      Value buf;
+      auto valType = acc[i].getType();
+
+      // row_shr:8 bank_mask:0xc
+      buf = createDppReduxOp(valType, acc[i], 8 + 0x110, 0xF, 0xC, true);
```

Comment:
Should we define some symbols for the magic number like `0x110` like the `DppCtrl` enum in `llvm/lib/Target/AMDGPU/SIDefines.h`?